### PR TITLE
chore: add limit checking to relay message

### PR DESCRIPTION
**Pull request by quake**

### CODE REVIEW

File: sync/src/relayer/block_proposal_process.rs [hunk 1]
```diff
@@ -16,12 +18,33 @@ pub enum Status {
...
+            peer,
+        }
     }
 
     pub fn execute(self) -> Result<Status, FailureError> {
         let snapshot = self.relayer.shared().snapshot();
+        {
+            let block_proposals = self.message;
+            if (block_proposals.transactions().len() as u64)
+                > snapshot.consensus().max_block_proposals_limit()
```
Comment (zhangsoledad) [inline sync/src/relayer/block_proposal_process.rs]: Should be 2 * snapshot.consensus().max_block_proposals_limit();

we send request_proposal_txs in 2 place.
One is `CompactBlockProcess`, we request_proposal_txs for one block,
Another one  is `BlockTransactionsProcess`, we request_proposal_txs for uncles.
we might request two uncles's proposal (the worst case, all fresh).

check https://github.com/nervosnetwork/ckb/blob/develop/sync/src/relayer/block_transactions_process.rs#L114-L124

File: sync/src/relayer/get_block_proposal_process.rs [hunk 2]
```diff
@@ -28,6 +28,20 @@ impl<'a> GetBlockProposalProcess<'a> {
     }
 
     pub fn execute(self) -> Result<(), FailureError> {
+        let snapshot = self.relayer.shared.snapshot();
+        {
+            let get_block_proposal = self.message;
+            if (get_block_proposal.proposals().len() as u64)
+                > snapshot.consensus().max_block_proposals_limit()
```
Comment (zhangsoledad) [inline sync/src/relayer/get_block_proposal_process.rs]: Same as above

File: sync/src/relayer/block_proposal_process.rs (hunk 1 above)
Comment (quake) [reply sync/src/relayer/block_proposal_process.rs]: Thanks for pointing out, I never noticed this issue before and think that we should increase `MAX_FRAME_LENGTH_RELAY` to 4MB for extreme cases:

https://github.com/nervosnetwork/ckb/blob/c9ab6bdaaca237913869b66fd85ccb64ef5ab7ed/network/src/lib.rs#L39

Comment (quake) [reply sync/src/relayer/block_proposal_process.rs]: Updated, please review again, thanks.

File: sync/src/relayer/compact_block_process.rs [hunk 3]
```diff
@@ -61,6 +61,27 @@ impl<'a> CompactBlockProcess<'a> {
 
     pub fn execute(self) -> Result<Status, FailureError> {
         let snapshot = self.relayer.shared.snapshot();
+        {
+            let compact_block = self.message;
+            if compact_block.uncles().len() > snapshot.consensus().max_uncles_num() {
```
[the remark this answers is not in the document]
Comment (quake) [reply sync/src/relayer/compact_block_process.rs]: `UncleVerifier` is only used in chain block processing, here is the first level checking of relay message.

File: sync/src/relayer/block_proposal_process.rs [hunk 4]
```diff
@@ -16,12 +18,33 @@ pub enum Status {
...
+            peer,
+        }
     }
 
     pub fn execute(self) -> Result<Status, FailureError> {
         let snapshot = self.relayer.shared().snapshot();
+        {
+            let block_proposals = self.message;
+            let limit = snapshot.consensus().max_block_proposals_limit()
+                * (snapshot.consensus().max_uncles_num() as u64);
```
[the remark this answers is not in the document]
Comment (quake) [reply sync/src/relayer/block_proposal_process.rs]: > should be `max_block_proposals_limit * (max_uncles_num + 1)`, including the block itself.

please refer to @zhangsoledad 's comment

**Comment by doitian** [comment]
@keroro520 @zhangsoledad please review again.

File: sync/src/relayer/get_transactions_process.rs [hunk 5]
```diff
@@ -69,19 +81,40 @@ impl<'a> GetTransactionsProcess<'a> {
...
-            if let Err(err) = self.nc.send_message_to(self.peer, data) {
-                debug_target!(
-                    crate::LOG_TARGET_RELAY,
-                    "relayer send Transactions error: {:?}",
-                    err,
-                );
+            let mut relay_bytes = 0;
+            let mut relay_txs = Vec::new();
+            for tx in transactions {
+                if relay_bytes + tx.total_size() > MAX_RELAY_TXS_BYTES_PER_BATCH {
```
Comment (doitian) [inline sync/src/relayer/get_transactions_process.rs]: Will `relay_bytes + tx.total_size()` overflow?

Comment (quake) [reply sync/src/relayer/get_transactions_process.rs]: > Will `relay_bytes + tx.total_size()` overflow?

No, all txs are verified, their serialized size can't exceed network protocol frame max lenth

**Pull request merged**